crypto/schnorr.c

```diff
@@ -0,0 +1,252 @@
+/**
+ * Copyright (c) 2021 The Bitcoin ABC developers
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "schnorr.h"
+#include "hmac_drbg.h"
+#include "memzero.h"
+#include "rfc6979.h"
+
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+int jacobi(const bignum256 *_n, const bignum256 *_k) {
+  int t = 0;
+  int k_is_one = 0;
+  uint32_t r = 0;
+  bignum256 n = {0}, k = {0}, temp = {0};
+
+  assert(!bn_is_zero(_k) && bn_is_odd(_k));
+
+  bn_copy(_n, &n);
+  bn_copy(_k, &k);
+
+  bn_mod(&n, &k);
```

Comment:
Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this `bn_mod` too, but in fact this `bn_mod` seems redundant, since `n < k` is guaranteed given that `_n` is always a point coordinate and `_k` is the order of the finite field. I'd just add `bn_is_less(_n, _k)` to the `assert` above.

Reply:
> Same comment as https://github.com/trezor/trezor-firmware/pull/1509/files#r588659565 applies to this bn_mod too, but in fact this bn_mod seems redundant, since n < k is guaranteed given that _n is always a point coordinate and _k is the order of the finite field.

Yes, this is because the coordinates of a point that is an output of `scalar_multiply`, `point_multiply` or `point_add` are fully reduced.

> I'd just add bn_is_less(_n, _k) to the assert above.

I agree.

Reply:
Actually I take back my comment about adding the assert, because the algorithm works correctly even if the assertion doesn't hold, so it would be misleading to have it there. I just removed the `bn_mod()` in https://github.com/trezor/trezor-firmware/pull/1509/commits/2bf51f3a2c2b6e39e88f3face3e9a3be67a634ee.